### PR TITLE
[statsd] Add Metrics and mocking

### PR DESCRIPTION
**Pull request by ariens-shopify**
### Description
This adds statsd related functionality and testing capabilities.

Pulls in the statsd related work that was added to Shopify/catwalk-server in https://github.com/Shopify/catwalk-server/pull/15  to Goose.

### CODE REVIEW

File: statsd/statsd.go [hunk 1]
```diff
@@ -43,6 +46,26 @@ func SetBackend(b Backend) {
...
 
+// ErrUnknownBackend is returned when the statsd backend implementation is not known.
+var ErrUnknownBackend = fmt.Errorf("unknown statsd backend type")
+
+// SetBackendFor returns the appropriate Backend for the given implementation and host.
+func SetBackendFor(impl, addr, prefix string, tags ...string) error {
+	var err error
+	switch strings.ToLower(impl) {
+	case "datadog":
+		currentBackend, err = NewDatadogBackend(addr, prefix, tags)
```
Comment (lavoiesl) [inline statsd/statsd.go]: I don't think this should override the currentBackend if there's an error.

File: statsd/statsd.go [hunk 2]
```diff
@@ -43,6 +46,26 @@ func SetBackend(b Backend) {
 	currentBackend = b
 }
 
+// ErrUnknownBackend is returned when the statsd backend implementation is not known.
+var ErrUnknownBackend = fmt.Errorf("unknown statsd backend type")
+
+// SetBackendFor returns the appropriate Backend for the given implementation and host.
+func SetBackendFor(impl, addr, prefix string, tags ...string) error {
```
Comment (lavoiesl) [inline statsd/statsd.go]: How about decoupling the "Set" and add a `NewBackend(impl, addr, prefix string, tags ...string) (Backend, error)`?

**Comment by ariens-shopify** [comment]
Just a word of warning, this is going to be pretty WIP while I copy pasta from catwalk-server.  Once it's all in place I'll ping for feedback.

**Comment by ariens-shopify** [comment]
> Clarifying that in the function comments might help

I've clarified that the stats are based off the metrics name, tags, and rate.  HTH.

File: statsd/statsd.go [hunk 3]
```diff
@@ -43,6 +46,27 @@ func SetBackend(b Backend) {
...
+		b, err = NewDatadogBackend(addr, prefix, tags)
+	case "log":
+		b = NewLogBackend(prefix, tags)
+	case "null":
+		b = NewNullBackend()
+	default:
+		return nil, errors.WithStack(ErrUnknownBackend)
+	}
+
+	return b, err
```
Comment (pseudomuto) [inline statsd/statsd.go]: The error in the default case is wrapped via `WithStack` (a good thing), seems like this should be done here as well. FWIW, `errors.Wrap(err, "...")` returns `nil`, when `err` is `nil`.

**Pull request merged**